server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -327,6 +333,23 @@ private void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collecto
         // catch early terminated exception and rethrow?
         Bits liveDocs = ctx.reader().getLiveDocs();
         BitSet liveDocsBitSet = getSparseBitSetOrNull(liveDocs);
+        if (isApproximateableRangeQuery()) {
+            ApproximateScoreQuery query = ((ApproximateScoreQuery) ((IndexOrDocValuesQuery) searchContext.query()).getIndexQuery());
+            ((ApproximatePointRangeQuery) query.getApproximationQuery()).setSize(
+                Math.max(searchContext.size(), searchContext.trackTotalHitsUpTo())
+            );
+            if (searchContext.request() != null && searchContext.request().source() != null) {
+                FieldSortBuilder primarySortField = FieldSortBuilder.getPrimaryFieldSortOrNull(searchContext.request().source());
+                if (primarySortField != null
+                    && primarySortField.missing() == null
+                    && primarySortField.getFieldName().equals(((RangeQueryBuilder) searchContext.request().source().query()).fieldName())) {
+                    if (primarySortField.order() == SortOrder.DESC) {
+                        ((ApproximatePointRangeQuery) query.getApproximationQuery()).setSortOrder(SortOrder.DESC);
```

Comment:
I'm thinking this logic doesn't belong in `ContextIndexSearcher`, since it's specific to the range approximation.

To generalize the idea of query approximation across many query types, I'm thinking the following might work:

1. Define an `ApproximatableQuery` abstract subclass of `Query` that `ApproximatePointRangeQuery` will extend. It should have a method that takes a `SearchContext` and returns true if the approximation is applicable (since every query will have its own context-specific logic).
2. ApproximateScoreQuery should have a method like `setContext(SearchContext)`. It will pass the `SearchContext` to its `ApproximatableQuery` to ask if the approximation will work. This decides which query it delegates to. (IMO, if `setContext()` isn't called, it shouldn't delegate and should throw an `IllegalStateException`.)
3. Here in `ContextIndexSearcher`, check if the top-level query is `instanceof ApproximateScoreQuery`, and if so, call `setContext()`.
4. For the specific case of `IndexOrDocValuesQuery`, we should have an `ApproximatableIndexOrDocValuesQuery` that holds the "real" index query, the approximate index query, and the doc values query (since I don't *think* DV queries will ever be approximatable).
5. Extending this logic in future means defining new `ApproximatableQuery` types and returning `ApproximateScoreQuery` from the appropriate `FieldType`s' query methods.

What do you think?